src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1531 @@
+
+        public IList<object> GetMembersByInferredType(PSTypeName typename, bool isStatic, Func<object, bool> filter)
+        {
+            List<object> results = new List<object>();
+
+            Func<object, bool> filterToCall = filter;
+            if (typename.Type != null)
+            {
+                AddMembersByInferredTypesClrType(typename, isStatic, filter, filterToCall, results);
+            }
+            else if (typename.TypeDefinitionAst != null)
+            {
+                AddMembersByInferredTypeDefinitionAst(typename, isStatic, filter, filterToCall, results);
+            }
+            else
+            {
+                // Look in the type table first.
+                if (!isStatic)
+                {
+                    var consolidatedString = new ConsolidatedString(new[] { typename.Name });
+                    results.AddRange(ExecutionContext.TypeTable.GetMembers<PSMemberInfo>(consolidatedString));
+                }
+
+                AddMembersByInferredTypeCimType(typename, results, filterToCall);
+            }
+
+            return results;
+        }
+
+        private void AddMembersByInferredTypeCimType(PSTypeName typename, List<object> results, Func<object, bool> filterToCall)
+        {
+            string cimNamespace;
+            string className;
+            if (ParseCimCommandsTypeName(typename, out cimNamespace, out className))
+            {
+                AddCommandWithPreferenceSetting(_powerShell, "CimCmdlets\\Get-CimClass")
+                    .AddParameter("Namespace", cimNamespace)
+                    .AddParameter("Class", className);
+                Exception unused;
+                var classes = Helper.ExecuteCurrentPowerShell(out unused);
+                foreach (var cimClass in classes.Select(PSObject.Base).OfType<CimClass>())
+                {
+                    results.AddRange(filterToCall != null
+                        ? cimClass.CimClassProperties.Where(filterToCall)
+                        : cimClass.CimClassProperties);
+                }
+            }
+        }
+
+        private void AddMembersByInferredTypeDefinitionAst(PSTypeName typename, bool isStatic, Func<object, bool> filter, Func<object, bool> filterToCall,
+            List<object> results)
+        {
+            if (CurrentTypeDefinitionAst != typename.TypeDefinitionAst)
+            {
+                if (filterToCall == null)
+                    filterToCall = o => !IsMemberHidden(o);
+                else
+                    filterToCall = o => !IsMemberHidden(o) && filter(o);
+            }
+
+            bool foundConstructor = false;
+            foreach (var member in typename.TypeDefinitionAst.Members)
+            {
+                bool add = false;
+                var propertyMember = member as PropertyMemberAst;
+                if (propertyMember != null)
+                {
+                    if (propertyMember.IsStatic == isStatic)
+                    {
```

Comment:
This can be simplified to `add = propertymember.IsStatic == isStatic`, with a similar change in the else branch, then the default initialization isn't necessary anymore.